Document has_many through association with source_type using associated model within the current module's namespace

monorkin opened the pull request:
### Motivation / Background

Defining `source_type` can cause an association to fetch records from a
different model than the one defined in `source_type`.

For example, if there are three models - `User`, `Device` and `User::Device`.
Defining a `has_many :favorite_devices, through: ...` relationship on
the `User` that uses `source_type: "Device"` should return objects
of class `Device`, but it returns objects of class `User::Device` instead.

This behavior can be corrected by specifying `class_name: "::Device"` on the
`has_many :favorite_devices, through: ...` relationship. With it the
association returns objects of type `Device`.

I believe that this is a bug because the behavior changes based on the model
that resolves the association. E.g. if I add a fourth model - `Widget` with
a `has_many :favorite_devices, through: :user` then this association returns
`Device` objects.

<details>
  <summary>Test case</summary>

  ```ruby
  # frozen_string_literal: true

  require "bundler/inline"

  gemfile(true) do
    source "https://rubygems.org"

    git_source(:github) { |repo| "https://github.com/#{repo}.git" }

    rails_options = if ENV.key?("LOCAL")
      { path: "#{Dir.pwd}/rails" }
    else
      { github: "rails/rails", branch: "main" }
    end

    gem "rails", rails_options
    gem "sqlite3"
    gem "timeout" # Fixes an issue with Ruby 3.1, ignore this
  end

  require "active_record"
  require "minitest/autorun"
  require "logger"

  # This connection will do for database-independent bug reports.
  ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
  ActiveRecord::Base.logger = Logger.new(STDOUT)

  ActiveRecord::Schema.define do
    create_table :users, force: true do |t|
    end

    create_table :user_devices, force: true do |t|
      t.integer :user_id
      # t.string :push_notifification_token
      # t.string :locale
    end

    create_table :widgets, force: true do |t|
      t.integer :user_id
      # t.string :name
    end

    create_table :user_favorites, force: true do |t|
      t.integer :user_id
      t.string :favorable_type
      t.integer :favorable_id
    end

    create_table :devices, force: true do |t|
      # t.string :name
    end
  end

  # A person that uses the app
  class User < ActiveRecord::Base
    has_many :devices,
             class_name: "User::Device"
    has_many :favorites,
             class_name: "User::Favorite"
    has_many :favorite_devices,
             through: :favorites,
             source: :favorable,
             source_type: "Device"
    has_many :widgets
  end

  # Anything the user wants to pin as their favorite
  # E.g. to show up in a dashboard after they sign in
  class User::Favorite < ActiveRecord::Base
    belongs_to :favorable, polymorphic: true
    belongs_to :user
  end

  # Used for session tracking and push notification
  # This model represents a person's phone or browser and would usually store
  # their push notification token, preffered language for that device and other
  # metadata
  class User::Device < ActiveRecord::Base
    belongs_to :user
  end

  # An IoT devices
  # THis model stores all configration data for the device and is used to
  # associate other models to a device
  class Device < ActiveRecord::Base
  end

  # An UI widget
  class Widget < ActiveRecord::Base
    belongs_to :user
    has_many :favorite_devices, through: :user
  end

  class BugTest < Minitest::Test
    def test_association_stuff
      user = User.create!
      widget = user.widgets.create!

      # The ID is important to demo this problem, that's why I set it manually
      # If the ID of the User::Device record and the Device record are the same
      # then this test passes due to a false positive
      user_device = User::Device.create!(id: 13, user: user)
      washing_machine = Device.create!(id: 13)
      vacuum_cleaner = Device.create!(id: 12)

      # The user adds an IoT Device to their favorites
      favorite1 = User::Favorite.create!(user: user, favorable: washing_machine)
      favorite2 = User::Favorite.create!(user: user, favorable: vacuum_cleaner)

      assert_equal [user_device], user.devices,
        "The user can access their devices (phones / browsers)"
      assert_equal [favorite1, favorite2], user.favorites,
        "The user can access their favorites"

      assert_equal ["Device", "Device"], user.favorites.pluck(:favorable_type),
        "The user's favorites are models of class Device"
      assert_equal [washing_machine.id, vacuum_cleaner.id].sort,
        user.favorites.pluck(:favorable_id).sort,
        "The user's favorites have the same IDs as the associated devices"

      assert_equal [washing_machine, vacuum_cleaner],
        [favorite1.favorable, favorite2.favorable]
        "The user can access their favorite device through the User::Favorite object"
      assert_equal [washing_machine, vacuum_cleaner].sort_by(&:id),
        widget.favorite_devices.sort_by(&:id),
        "The widget can access the user's favorite devices"

      # These fail because `user.favorite_devices` queries User::Device instead
      # of Device when `#compute_class` is called
      assert_equal [washing_machine, vacuum_cleaner].sort_by(&:id),
        user.favorite_devices.sort_by(&:id),
        "The user can access their own favorite devices"
      assert_equal [washing_machine.id, vacuum_cleaner.id].sort,
        user.favorite_device_ids.sort,
        "The user can access the IDs of their own favorite devices"
    end
  end
  ```
</details>

### Detail

When you specify a `source_type` it gets used in the `klass` method which
is responsible for returning the class of the objects returned by the
association.

The `klass` method determines the class using two other methods - `class_name`
and `compute_class`.

`class_name` returns the sanitized value from the `derive_class_name` method,
which returns the value from the passed `source_type` option.

`compute_class` tries to resolve this class name to a class in the usual way.
In the example from before, where the `User` has an association with
`source_type: "Device"`, it would search for the `User::Device` class first and
then for the `Device` class. Since `User::Device` exists it gets used as the
association's model instead of `Device`.

When a `Widget` calls `favorite_devices` through its
`has_many :favorite_devices, through: :user` association, `compute_class`
returns the `Device` class because it can't find a `Widget::Device` class and
falls back to the `Device` class.

To fix this, I made the returned class name from `derive_class_name`
absolute by prefixing it with `::` if the value comes from the `source_type`
option.

This makes class lookups from `source_type` consistent with the query behavior,
resolves similar issues with mixed-in and inherited associations that use
`source_type`, and it makes the class name in the polymorphic `*_type` column
absolute when used with `source_type` (which seems to be how polymorphic
relationships are supposed to be stored).

The downside of this solution is that it breaks associations that relied on
this lookup behavior for `source_type`.

### Additional information

This patch still allows the use of `class_name` with `source_type` and it works as
usual - `class_name` takes precedence. Maybe the combined use of `source_type`
and `class_name` should prohibited? I haven't investigated this possibility.

---

I'm not sure what the convention for adding new test models is.

I tried extending the `Tag` test model to have an associated `Rating` to avoid
adding three extra models, but that caused other unrelated tests to fail so I
opted for the extra few models instead.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rafaelfranca commented:
I'm not so sure this is a bug. Even if it was, this is now changing an existing behavior of Rails that will probably break some application that rely on this behavior.

This also happens with the `class_name` option.

In summary I'm more inclined to make this behavior documented and tested than changing it.

monorkin commented:
Hi Rafael, 
thank you for the comment.

I see your point. Would you like me to update this PR to document this behavior and change the tests to check for it or should I open a separate PR?

monorkin committed:
Document namespace lookup behaviour of source_type

source_type will look for the associated model's class within the
current modules namespace. This is undocumented and untested behaviour.
This commit adds a test that checks this behaviour, and documents it in
the guides.

monorkin commented:
Hey @rafaelfranca I've updated this PR to add [a test for this lookup behavior](https://github.com/rails/rails/pull/46655/files#diff-b50a42dc13373c1657aa3dac30cd0a3d94c5532bb577354ba4d5386adbb9175aR384) and also updated [the guides for has_one/has_many to explain the lookup behavior with an example](https://github.com/rails/rails/pull/46655/files#diff-d977574b785ce95fe656c58211c668992bf95214d52d3ffadb409ef8f69ad287).
